File: 0086. Partition List.java
// Medium
// LinkedList
// Time:O(n), Space:O(1)
// https://leetcode.cn/problems/partition-list/

class Solution {
    public ListNode partition(ListNode head, int x) {
        ListNode smallDummy = new ListNode(-1), bigDummy = new ListNode(-1);
        ListNode smaller = smallDummy, bigger = bigDummy;
        while (head != null) {
            if (head.val < x) {
                smaller.next = head;
                smaller = smaller.next;
            } else {
                bigger.next = head;
                bigger = bigger.next;
            }
            head = head.next;
        }
        bigger.next = null;
        smaller.next = bigDummy.next;
        return smallDummy.next;
    }
}

File: 0001-0500/0086. Partition List.go
package main
// Medium
// Time:O(n), Space:O(1)
// https://leetcode.cn/problems/partition-list/

func partition(head *ListNode, x int) *ListNode {
    lessDummy := &ListNode{}
    greaterDummy := &ListNode{}

    less := lessDummy
    greater := greaterDummy
    cur := head

    for cur != nil {
        if cur.Val < x {
            less.Next = cur
            less = less.Next
        } else {
            greater.Next = cur
            greater = greater.Next
        }
        cur = cur.Next
    }
    greater.Next = nil
    less.Next = greaterDummy.Next
    return lessDummy.Next
}
